src/connection.js

```diff
@@ -1274,6 +1306,95 @@ class Connection extends EventEmitter {
     this.transitionTo(this.STATE.SENT_LOGIN7_WITH_STANDARD_LOGIN);
   }
 
+  getRequestOptions() {
+    var tokenHost = '169.254.169.254';
+    var tokenEndpoint = '/metadata/identity/oauth2/token';
+    var clinetIdKeyName = '&client_id=';
+    const apiVersionKeyName = '&api-version=';
+    const clinetId = this.config.authentication.options.MSIclientID;
+    const msiEndpoint = process.env.MSI_ENDPOINT;
+    const msiSecret = process.env.MSI_SECRET;
+    var tokenPath = '';
+    var tokenPort = '';
+    var tokenHeaders = { 'Metadata': true };
+    // Api version supported on Azure VM
+    var apiVersion = '2018-02-01';
+
+    // If MSI_ENDPOINT and MSI_SECRET are defined, used these instead of the hard-coded host
+    if (undefined !== msiEndpoint && undefined !== msiSecret) {
+      // Parse the environment variable as a url and set corresponding request components
+      var appUrl = url.parse(msiEndpoint, true);
+      tokenHost = appUrl.hostname;
+      tokenPath = appUrl.pathname;
+      tokenPort = appUrl.port;
+      tokenHeaders = {
+        'Secret': msiSecret
+      };
+      // Api version supported on Azure App
+      apiVersion = '2017-09-01';
+      clinetIdKeyName = '&clientid=';
+    } else {
+      tokenPath = tokenEndpoint;
+    }
+
+    tokenPath += '?resource=' + this.fedAuthInfoToken.spn;
+
+    // If connection config contians a client id
+    if ('' !== clinetId) {
+      tokenPath += clinetIdKeyName + clinetId;
+    }
+
+    tokenPath += apiVersionKeyName + apiVersion;
+
+    // Append the API version for connection within an Azure VM
+    var options = {
+      host: tokenHost,
+      path: tokenPath,
+      port: tokenPort,
+      method: 'GET',
+      headers: tokenHeaders
+    };
+    return options;
+  }
+
+  getMSIToken() {
+    // Get the token from the azure source
+    http.get(this.getRequestOptions(), (tokenReponse) => {
+      let data = '';
+
+      // Keep getting the packages
+      tokenReponse.on('data', (chunk) => {
+        data += chunk;
+      });
+
+      // Done getting the packages
+      tokenReponse.on('end', () => {
+        // If the response data contains an error or empty
+        if (-1 !== data.indexOf('error') || '' === data) {
```

Comment:
In cases where the MSI endpoint returns an error, isn't there a proper JSON response returned? Checking the response for an the word `error` does not seem to be a very stable approach (just imagine the same characters randomly turning up in the token or similar).

Reply:
Agreed, should think deeper on this.  By the way, can we expose the error message in detail or we can just use a general error message to hide the original error message for security reasons?